subsys/net/ip/tcp2.c

```diff
@@ -480,89 +631,43 @@ static size_t tcp_data_get(struct tcp *conn, struct net_pkt *pkt)
 	return len;
 }
 
-static int tcp_finalize_pkt(struct net_pkt *pkt)
-{
-	net_pkt_cursor_init(pkt);
-
-	if (IS_ENABLED(CONFIG_NET_IPV4) && net_pkt_family(pkt) == AF_INET) {
-		return net_ipv4_finalize(pkt, IPPROTO_TCP);
-	}
-
-	if (IS_ENABLED(CONFIG_NET_IPV6) && net_pkt_family(pkt) == AF_INET6) {
-		return net_ipv6_finalize(pkt, IPPROTO_TCP);
-	}
-
-	return -EINVAL;
-}
-
-static int tcp_header_add(struct tcp *conn, struct net_pkt *pkt, u8_t flags)
-{
-	NET_PKT_DATA_ACCESS_DEFINE(tcp_access, struct tcphdr);
-	struct tcphdr *th;
-
-	th = (struct tcphdr *)net_pkt_get_data(pkt, &tcp_access);
-	if (!th) {
-		return -ENOBUFS;
-	}
-
-	memset(th, 0, sizeof(struct tcphdr));
-
-	th->th_sport = conn->src.sin.sin_port;
-	th->th_dport = conn->dst.sin.sin_port;
-
-	th->th_off = 5;
-	th->th_flags = flags;
-	th->th_win = htons(conn->win);
-	th->th_seq = htonl(conn->seq);
-
-	if (ACK & flags) {
-		th->th_ack = htonl(conn->ack);
-	}
-
-	return net_pkt_set_data(pkt, &tcp_access);
-}
-
-static int ip_header_add(struct tcp *conn, struct net_pkt *pkt)
-{
-	if (IS_ENABLED(CONFIG_NET_IPV4) && net_pkt_family(pkt) == AF_INET) {
-		return net_context_create_ipv4_new(conn->context, pkt,
-						&conn->src.sin.sin_addr,
-						&conn->dst.sin.sin_addr);
-	}
-
-	if (IS_ENABLED(CONFIG_NET_IPV6) && net_pkt_family(pkt) == AF_INET6) {
-		return net_context_create_ipv6_new(conn->context, pkt,
-						&conn->src.sin6.sin6_addr,
-						&conn->dst.sin6.sin6_addr);
-	}
-
-	return -EINVAL;
-}
-
 static void tcp_out(struct tcp *conn, u8_t flags, ...)
 {
 	struct net_pkt *pkt;
 	size_t len = 0;
 	int r;
 
-	pkt = tcp_pkt_alloc(conn, sizeof(struct tcphdr));
-	if (!pkt) {
-		goto fail;
-	}
-
 	if (PSH & flags) {
-		struct net_pkt *data_pkt;
+		/* How the data is send:
+		 * - if there is no pending data yet, allocate net_pkt
+		 * - append data to existing list of pending data
+		 * - trigger send
+		 * - the send function will figure out how much data to
+		 *   send according to send window size and mss
+		 */
 		va_list ap;
```

Comment:
Could you take the opportunity to remove the usage of variable argument list on this function? It would take a net_pkt pointer as argument after flags, and that's it. It's way more readable that way.

Reply:
That is the ultimate intention, the va_list is not really needed here. But I would like to avoid changing it in this PR in order to minimize the changes.

Reply:
ok